Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/distance.h

```diff
@@ -227,6 +378,313 @@ sample_triangles(const FaceRange& triangles,
+    NamedParameters,
+    typename TriangleRange::const_iterator,
+    Random_points_in_triangle_soup<PointRange, typename TriangleRange::value_type, Creator>,
+    Creator,
+    Triangle_structure_sampler_for_triangle_soup<PointRange,
+      TriangleRange,
+      OutputIterator,
+      Geom_traits,
+      Creator,
+      NamedParameters>
+    >
+{
+  typedef typename TriangleRange::value_type TriangleType;
+  typedef Triangle_structure_sampler_for_triangle_soup<PointRange,
+  TriangleRange,
+  OutputIterator,
+  Geom_traits,
+  Creator,
+  NamedParameters> This;
+
+  typedef Triangle_structure_sampler_base<
+  OutputIterator,
+  Geom_traits,
+  NamedParameters,
+  typename TriangleRange::const_iterator,
+  Random_points_in_triangle_soup<PointRange, TriangleType, Creator>,
+  Creator,
+  This> Base;
+
+
+
+  typedef Random_points_in_triangle_soup<PointRange, TriangleType, Creator> Randomizer;
+  typedef typename TriangleRange::const_iterator TriangleIterator;
+
+  double min_edge_length_;
+  const TriangleRange& triangles;
+  const PointRange& points;
+  Triangle_structure_sampler_for_triangle_soup(const PointRange& pts,
+                                               const TriangleRange& trs,
+                                               OutputIterator& out,
+                                               NamedParameters np)
+    :Base(out, np),
+      points(pts), triangles(trs)
+  {
+    using parameters::choose_parameter;
+    using parameters::get_parameter;
+
+    min_edge_length_ = (std::numeric_limits<double>::max)();
+  }
+
+  std::pair<TriangleIterator,TriangleIterator> get_range()
+  {
+    return std::make_pair(triangles.begin(), triangles.end());
+  }
+  void sample_points()
+  {
+    this->out = std::copy(
+          points.begin(),
+          points.end(),
+          this->out);
+  }
+  double get_minimum_edge_length()
+  {
+    if(min_edge_length_ != (std::numeric_limits<double>::max)())
+      return min_edge_length_;
+    for(auto tr : triangles)
+    {
+      for(std::size_t i = 0; i< 3; ++i)
+      {
+        typename Geom_traits::Point_3  a(points[tr[i]]),
+            b(points[tr[(i+1)%3]]);
+
+        double el = std::sqrt(
+              to_double( typename Geom_traits::Compute_squared_distance_3()(
+                           a, b)));
+        if (el > 0 && el < min_edge_length_)
+          min_edge_length_ = el;
+      }
+    }
+    return min_edge_length_;
+  }
+  template<typename Tr>
+  double get_tr_area(const Tr& tr)
+  {
+    return to_double(approximate_sqrt(
+                       this->geomtraits.compute_squared_area_3_object()(
+                         points[tr[0]],points[tr[1]], points[tr[2]])));
+  }
+
+  template<typename Tr>
+  void get_tr_points(const Tr& tr, typename Geom_traits::Point_3 points[])
+  {
+    for(int i=0; i<3; ++i)
+    {
+      points[i] = this->points[tr[i] ];
+    }
+  }
+  void ms_edges_sample(std::size_t ,
+                       double )
+  {
+    //don't sample edges in soup.
```

Comment:
Why is that?

Reply:
Because a soup does not have edges, so why sample them ?

Reply:
A triangle has 3 edges last time I checked. The only thing is whether you want to have to deal with the same edge being sampled potentially multiple times and how would you handle that case. Let's say it's not worth the effort for now.